Use schema as source of truth + support ones_like/empty_like

janeyx99 opened the pull request:
This change does 2 important things:
(a) Instead of relying on IValue type as source of truth, we use the schema as the source of truth, which is important as IValue types are overloaded and can ambiguously convert incorrectly. For example, a MemoryFormat will look like an int + get converted to an int64_t vs a MemoryFormat!

(b) This PR expands support for many more types to encompass way more schemas, e.g., Optional, Device, dtype, etc. The main win from this PR is the ability for aoti_torch_call_dispatcher to call TensorFactory ops like ones_like/empty_like!

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #149230
* __->__ #149052

janeyx99 committed:
[WIP] support ones_like with call_dispatcher shim API

[ghstack-poisoned]

janeyx99 committed:
Update on "[WIP] support ones_like with call_dispatcher shim API"




[ghstack-poisoned]

janeyx99 committed:
Update on "Use schema as source of truth + support ones_like/empty_like"




[ghstack-poisoned]

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1298,6 +1298,127 @@ AOTITorchError aoti_torch_zero_(AtenTensorHandle tensor) {
+      // return from<std::optional<inner_type::t>>(ivalue.toInnerTypeT()));
+      //
+      // BUT we do NOT have that type inner_type::t readily available, so we
+      // will manually unwrap and recursively call, similar to the
+      // from<std::optional<T>> function in torch/csrc/stable/library.h
```
```suggestion
      // will manually unwrap and recursively call, this implementation MUST be kept in sync with
      // from<std::optional<T>> function in torch/csrc/stable/library.h
```

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1298,6 +1298,127 @@ AOTITorchError aoti_torch_zero_(AtenTensorHandle tensor) {
+    const StableIValue stable_ivalue) {
+  switch (type->kind()) {
+    case c10::TypeKind::TensorType: {
+      // stable_ivalue must be an ATH
```
not sure that comment is helpful

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1298,6 +1298,127 @@ AOTITorchError aoti_torch_zero_(AtenTensorHandle tensor) {
+      // return c10::IValue(to<std::optional<inner_type::t>>(stable_ivalue));
+      //
+      // BUT we do NOT have that type inner_type::t readily available, so we
+      // will manually unwrap and recursively call, similar to the to<T>
```
similar: MUST be kept in sync.

albanD reviewed (commented):
Please add the table of type correspondances.
Inline in C++ or in docs/notes.

janeyx99 committed:
Update on "Use schema as source of truth + support ones_like/empty_like"

This change does 2 important things:
(a) Instead of relying on IValue type as source of truth, we use the schema as the source of truth, which is important as IValue types are overloaded and can ambiguously convert incorrectly. For example, a MemoryFormat will look like an int + get converted to an int64_t vs a MemoryFormat!

(b) This PR expands support for many more types to encompass way more schemas, e.g., Optional, Device, dtype, etc. The main win from this PR is the ability for aoti_torch_call_dispatcher to call TensorFactory ops like ones_like/empty_like!




[ghstack-poisoned]

albanD commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,48 @@
+|  type in custom extension    |   StableIValue representation   |   type in libtorch  |   IValue  | 
+| -------- | ------- | ------- | ------- |
+| std::optional\<S> | \*reinterpret_cast\<(StableIValue\*)\*>, pointer to a StableIValue recursively defined | std::optional\<T> | OptionalType |
+| std::nullopt | \*reinterpret_cast\<nullptr_t\*> | IValue() | NoneType |
```
not correct anymore.

janeyx99 replied:
no this one is correct?

zou3519 replied:
Umm schemas don't exactly have NoneType in them

janeyx99 replied:
I changed the table so it uses the schema types like in native_functions.yaml so things should be clearer now

janeyx99 committed:
Update on "Use schema as source of truth + support ones_like/empty_like"

This change does 2 important things:
(a) Instead of relying on IValue type as source of truth, we use the schema as the source of truth, which is important as IValue types are overloaded and can ambiguously convert incorrectly. For example, a MemoryFormat will look like an int + get converted to an int64_t vs a MemoryFormat!

(b) This PR expands support for many more types to encompass way more schemas, e.g., Optional, Device, dtype, etc. The main win from this PR is the ability for aoti_torch_call_dispatcher to call TensorFactory ops like ones_like/empty_like!




[ghstack-poisoned]

albanD commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,48 @@
+| int32_t | \*reinterpret_cast\<uint64_t\*> | at::MemoryFormat | MemoryFormatType |
+| bool | \*reinterpret_cast\<uint64_t\*> | bool | BoolType |
+| int64_t | \*reinterpret_cast\<uint64_t\*> | int64_t | IntType |
+| float | \*reinterpret_cast\<uint64_t\*> | double | FloatType |
```
Why not double on the extension side?!

albanD commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,48 @@
+3. type in libtorch: type used within libtorch.so.
+4. Schema Type: type as described by the schema, which we hail as the source of truth for both ATen ops in native_functions.yaml and for user defined custom ops registered to the dispatcher.
+
+|  type in custom extension    |   StableIValue representation   |   type in libtorch  |   Schema Type  |
```
The content of the Schema Type is not correct or at least doesn't match the types used in the schema strings.

janeyx99 replied:
Rewrote, removed these that aren't referenced in native_functions.yaml:

```
| ? | ? | torch::jit::Function* | Function |
| ? | ? | c10::RRefInterface | RRef(Type) |
| ? | ? | at::Quantizer | Quantizer |
| ? | ? | ivalue::EnumHolder | AnyEnumType |
| ? | ? | c10::Dict<IValue, IValue> | Dict(Type, Type) |
| ? | ? | torch::jit::Module | Module |
| ? | ? | ivalue::Object | ClassType / AnyClassType |
| ? | ? | c10::Capsule | Capsule |
| ? | ? | ivalue::PyObjectHolder | PyObjectType |
| ? | ? | ivalue::Future | Future(Type) |
| ? | ? | ivalue::Await | Await(Type) |
| ? | ? | ? | Any |
| ? | ? | ? | InterfaceType |
| ? | ? | ? | UnionType |
| ? | ? | ? | DynamicType |
| ? | ? | ? | VarType |
```
@zou3519 are these represented with torchbind in c++ + thus still worth adding back eventually?

zou3519 replied:
a lot of these are for TorchScript, just leave them off for now.

albanD commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,48 @@
+| bool | \*reinterpret_cast\<uint64_t\*> | bool | BoolType |
+| int64_t | \*reinterpret_cast\<uint64_t\*> | int64_t | IntType |
+| float | \*reinterpret_cast\<uint64_t\*> | double | FloatType |
+| ? | ? | c10::Device | DeviceObjType |
```
You did define the StableIValue for this one by side effect in this PR.
We actually define this for any object <64bits, should we say it?

janeyx99 commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,48 @@
+
+|  type in custom extension    |   StableIValue representation   |   type in libtorch  |   IValue  | 
+| -------- | ------- | ------- | ------- |
+| std::optional\<S> | \*reinterpret_cast\<(StableIValue\*)\*>, pointer to a StableIValue recursively defined | std::optional\<T> | OptionalType |
```
wait isn't this line 12

zou3519 commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,48 @@
+1. type in custom extension: type used within the end user custom library.
+2. StableIValue representation: a stable conversion of the type to liaison between the user model vs libtorch.so.
+3. LibTorch type: type used within libtorch.so.
+4. Schema Type: type as described by the schema, which we hail as the source of truth for both ATen ops in native_functions.yaml and for user defined custom ops registered to the dispatcher.
```
TORCH_LIBRAY -> TORCH_LIBRARY

Also, we should spell out "custom operators" in documentation, I know we don't always do that

janeyx99 replied:
done!

zou3519 commented on torch/csrc/stable/library.h:
```diff
@@ -4,23 +4,80 @@
+
+// Specialization for std::optional
 template <typename T>
+StableIValue from(std::optional<T> val) {
```
What is T? Does this need to be a reference instead?

janeyx99 replied:
no, we expect an rvalue to be passed in from the user

zou3519 commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,48 @@
+| -------- | ------- | ------- | ------- |
+| std::optional\<S> | \*reinterpret_cast\<(StableIValue\*)\*>, pointer to a StableIValue recursively defined | std::optional\<T> | OptionalType |
+| std::nullopt | \*reinterpret_cast\<nullptr_t\*> | IValue() | NoneType |
+| RAIIATH | \*reinterpret_cast\<uint64_t\*> of AtenTensorHandle | at::Tensor |  TensorType |
```
<img width="861" alt="image" src="https://github.com/user-attachments/assets/6a092e91-4fc1-4cf7-a7ae-95880fc18ec0" />
Are all the *'s intentional? I'm having a hard time parsing this

janeyx99 replied:
The *s are intentional--we're type punning and casting to avoid compiler warnings

zou3519 commented on torch/csrc/stable/library.h:
```diff
@@ -4,23 +4,80 @@
+template <
+    typename T,
+    std::enable_if_t<std::is_same_v<T, std::nullopt_t>, bool> = true>
+T to(StableIValue val) {
```
Just to check, StableIValue's representation is some 64-bit type?

janeyx99 replied:
it is using StableIValue = uint64_t

zou3519 commented on torch/csrc/stable/library.h:
```diff
@@ -4,23 +4,80 @@
+  delete sivp;
+
+  return std::make_optional(inner_val);
+}
```
The thing that is confusing to me that would benefit from more documentation is why we are deleting sivp (and what the memory management strategy for it is).

zou3519 reviewed (commented):
At a high level I like using the schema as the source of truth. For the implementation, the optional<T> stuff memory handling looked fishy. I didn't read the docs

janeyx99 committed:
Update on "Use schema as source of truth + support ones_like/empty_like"

This change does 2 important things:
(a) Instead of relying on IValue type as source of truth, we use the schema as the source of truth, which is important as IValue types are overloaded and can ambiguously convert incorrectly. For example, a MemoryFormat will look like an int + get converted to an int64_t vs a MemoryFormat!

(b) This PR expands support for many more types to encompass way more schemas, e.g., Optional, Device, dtype, etc. The main win from this PR is the ability for aoti_torch_call_dispatcher to call TensorFactory ops like ones_like/empty_like!




[ghstack-poisoned]

zou3519 commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,34 @@
+| ? | ? | at::Storage | Storage |
+| ? | ? | at::Generator | Generator |
+| ? | ? | c10::List\<T> | Type[] |
+| ? | ? | ivalue::Tuple\<T> | (Type, ...) |
```
Is Tuple a valid operator type? I think we usually only handle lists

janeyx99 replied:
They're valid as return types

zou3519 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/libtorch_agnostic_kernel.cpp:
```diff
@@ -147,3 +149,39 @@ STABLE_TORCH_LIBRARY_FRAGMENT(libtorch_agnostic, m) {
+  auto mf = aoti_torch_memory_format_contiguous_format();
+
+  stack[0] = from(t.release());
+  stack[1] = from(std::optional(t_dtype));    // dtype
```
Yeah I don't feel good about the `new/malloc` inside `from` for std::optional. std::optional itself is a stack-based type (it does not do heap allocations). The new/malloc call does a heap allocation. heap allocations can be slow.

Since the StableIValue stuff is a low-level API (and I expect that you'll want to provide a higher-level API around it?) then it should be as performant as we can make it. An alternative is to do this is something like the following
```
auto stable_dtype = from(t_dtype);
stack[1] = asOptional(&stable_dtype);
stack[2] = asOptional(nullptr);
...
```

janeyx99 replied:
For future reference: we discussed this offline and boiled down the difference to being that perf of heap allocations = worse than stack. However, there is no easy way to go the reverse direction from IValue to StableIValue to user type on the stack as the first function will return too quickly.

Either way, I benchmarked the difference https://github.com/pytorch/pytorch/pull/149445 and we agreed it was minute enough to go forward with the current strategy.

zou3519 commented on docs/source/notes/libtorch_stable_abi.md:
```diff
@@ -0,0 +1,34 @@
+| int32_t | \*reinterpret_cast\<uint64_t\*> | at::Layout | Layout |
+| int32_t | \*reinterpret_cast\<uint64_t\*> | at::MemoryFormat | MemoryFormat |
+| bool | \*reinterpret_cast\<uint64_t\*> | bool | bool |
+| int64_t | \*reinterpret_cast\<uint64_t\*> | int64_t | int |
```
The way I am reading this is that you took an int64_t val, did a `reinterpret_cast<uint64_t*>(val)`, and then dereferenced it to get a uint64_t.
 
1. As written this is UB. So it probably means that we do a `reinterpret_cast<uint64_t*>(&val)`
2. Can this just be `reinterpret_cast<uint64_t>(val)`?

janeyx99 replied:
Ah, in code it is the first, I had just failed to document correctly. Sorry! - will fix

janeyx99 replied:
2. no, reinterpret_cast cannot cast with floats, so it'll fail for a subset of these

swolchok commented on torch/csrc/stable/library.h:
```diff
@@ -4,23 +4,110 @@
   static_assert(
       sizeof(T) <= sizeof(StableIValue),
       "StableLibrary stack does not support parameter types larger than 64 bits.");
   return *reinterpret_cast<StableIValue*>(&val);
```
I believe it is UB to call this unless `T` is `StableIValue`

janeyx99 commented:
@pytorchbot cherry-pick --onto release/2.7 -c fixnewfeature
